src/fact_checker/fact_topology.rs

```diff
@@ -0,0 +1,113 @@
+//! Fact topology type and helpers.
+//!
+//! Ported from https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/cairo/bootloaders/fact_topology.py
+
+use std::collections::HashMap;
+
+use cairo_vm::{
+    types::builtin_name::BuiltinName,
+    vm::runners::cairo_pie::{BuiltinAdditionalData, CairoPie, PublicMemoryPage},
+};
+
+use super::error::FactCheckerError;
+
+pub const GPS_FACT_TOPOLOGY: &str = "gps_fact_topology";
+
+/// Flattened fact tree
+#[derive(Debug, Clone)]
+pub struct FactTopology {
+    /// List of pairs (n_pages, n_nodes)
+    pub tree_structure: Vec<usize>,
+    /// Page sizes (pages are leaf nodes)
+    pub page_sizes: Vec<usize>,
+}
+
+/// Returns the fact topology from the additional data of the output builtin.
+pub fn get_fact_topology(cairo_pie: &CairoPie, output_size: usize) -> Result<FactTopology, FactCheckerError> {
+    if let Some(BuiltinAdditionalData::Output(additional_data)) = cairo_pie.additional_data.0.get(&BuiltinName::output)
+    {
+        let tree_structure = match additional_data.attributes.get(GPS_FACT_TOPOLOGY) {
```

Comment:
are there any docs on this or some reference code?

Reply:
There are reference link to the original Python sources + I've added (copied) a more extensive docs in the fact info module

Reply:
Got it, thanks!